rpc/providers/web-socket-proxy.ts

```diff
@@ -0,0 +1,120 @@
+type ReconnectOptions = {
+  attempts?: number;
+  maxAttempts?: number;
+  delay?: number;
+};
+
+type WebSocketClientOptions = {
+  webSocketFactory: () => WebSocket;
+  reconnect?: ReconnectOptions;
+};
+
+export class WebSocketProxy implements Pick<WebSocket, "send" | "close"> {
+  onopen?: (e: WebSocketEventMap["open"]) => void;
+  onclose?: (e: WebSocketEventMap["close"]) => void;
+  onmessage?: (e: WebSocketEventMap["message"]) => void;
+  onerror?: (e: WebSocketEventMap["error"]) => void;
+
+  #webSocket!: WebSocket;
+  #isClosed = false;
+  #isReconnecting = false;
+
+  #reconnectTimeoutId?: number;
+  #reconnectAttempts = 0;
+  #reconnectMaxAttempts: number;
+  #reconnectDelay: number;
+
+  constructor(private readonly options: WebSocketClientOptions) {
+    this.#reconnectMaxAttempts = options.reconnect?.maxAttempts ?? 5;
+    this.#reconnectDelay = options.reconnect?.delay ?? 1000;
+
+    this.#connect();
+  }
+
+  get isOpen(): boolean {
+    return this.#webSocket.readyState === WebSocket.OPEN;
+  }
+
+  send(data: string | ArrayBufferLike | Blob | ArrayBufferView): void {
+    this.#webSocket.send(data);
+  }
+
+  close(code?: number, reason?: string): void {
+    clearTimeout(this.#reconnectTimeoutId);
+
+    this.#isClosed = true;
+
+    const webSocket = this.#webSocket;
+
+    webSocket.onopen = null;
+    webSocket.onclose = null;
+    webSocket.onmessage = null;
+    webSocket.onerror = null;
+
+    webSocket.close(code, reason);
+
+    if (this.#webSocket.readyState === WebSocket.CLOSED) {
+      this.onclose?.(new CloseEvent("close", { wasClean: true, code, reason }));
+    } else {
+      webSocket.onclose = () => {
+        this.onclose?.(new CloseEvent("close", { wasClean: true, code, reason }));
+        webSocket.onclose = null;
+      };
+    }
+  }
+
+  #connect() {
+    this.#isClosed = false;
+    this.#isReconnecting = false;
+    this.#webSocket = this.options.webSocketFactory();
+
+    this.#webSocket.onopen = this.#onWSOpen;
```

Comment:
In situations like this, we prefer `#onWsOpen`, as the `WS` abbreviation is considered a term unto itself (hence we camelize). Same thing for `onWSMessage`->`onWsMessage` and others.